### DISCUSSION TITLE
Proper Prettier config?

### DISCUSSION THREAD

**Discussion by Hans5958** [ScratchAddons/ScratchAddons#162] [Discussion]
Can we have a proper Prettier config instead just the maximum width? Don't know if you guys prefer spaces or tabs, or you guys prefer 2 or 4 spaces.

**Comment by NoobTracker** [comment]
I prefer two spaces because reasons.

**Comment by NoobTracker** [comment]
I think that @JeffaloBob prefers tabs.

**Comment by NoobTracker** [comment]
But I think that most users prefer spaces.
Fun fact: People who use spaces instead of tabs earn more money!

**Comment by Hans5958** [comment]
1 message pls

I prefer tabs, because...
a. People can adjust their indentation width by themselves.
b. Why confuse between using 2 spaces, or 4, or 1, or 3, or even 8, when I can use a single tab character for everything? 
c. It is smaller in size, a single tab character vs multiple space characters. (thought I won't matter with this one)
I am adapting with each repository preference, but if I'm all alone, tabs.

Yeah, I have seen that survey. https://stackoverflow.blog/2017/06/15/developers-use-spaces-make-money-use-tabs/

**Comment by NoobTracker** [comment]
Well, I guess you can use some programs you convert it to tabs.
And the file size ... that's a really bad point. Using a tab instead of two spaces probably makes the files 2-3% smaller ... yeah 100KB or 97KB, that's a huuuuuuuuuuge difference.
And one of the main reasons I use space is that that's what most IDEs use so i got used to it. And really, the tab key is in a really awkward position.

Anyway, spaces are just more common and they are already used so I guess there is no reason to change that.

**Comment by Hans5958** [comment]
> And one of the main reasons I use space is that that's what most IDEs use so i got used to it. And really, the tab key is in a really awkward position.

The tabs are in the awkward position? Even if I need to use spaces, I will still use the tab button, because the tabs will automatically adjust what indentation should be entered.

> Anyway, spaces are just more common and they are already used so I guess there is no reason to change that.

You can't tell me what to do! /s

**Comment by NoobTracker** [comment]
I'm pretty sure that prettier always replaces tabs with the given number of spaces and you can also reverse that so, if you really want, you can use tabs. I still prefer spaces.

**Comment by Hans5958** [comment]
> 
> 
> I'm pretty sure that prettier always replaces tabs with the given number of spaces and you can also reverse that so, if you really want, you can use tabs. I still prefer spaces.

Yes, but if it is stated on the config. If it isn't, it will fallback to the environment config. (I use tabs here)

**Comment by WorldLanguages** [comment]
Unespecified configs fallback to Prettier defaults.

**Comment by WorldLanguages** [comment]
And Prettier defaults are nice. I just changed the line width so that every addon script can have `export default async` etc... in a single line.

**Comment by tb148** [comment]
Prettier is "opinionated" so I don't think we should be discussing about it.

**Comment by Hans5958** [comment]
> Prettier is "opinionated" so I don't think we should be discussing about it.

Not if it's enforced through the formatting workflow, I guess.

> Unespecified configs fallback to Prettier defaults.
> And Prettier defaults are nice. I just changed the line width so that every addon script can have export default async etc... in a single line.

Yes, but after searching for the config by going up the file tree. This may causes unwanted issue for people that developing with different enviroments, like me. If you prefer the defaults, then consider adding the default values then.

**Comment by WorldLanguages** [comment]
> Yes, but after searching for the config by going up the file tree. This may causes unwanted issue for people that developing with different enviroments, like me.

Are you sure that's how it works? The Prettier documentation doesn't seem to say anything about that. It explicitly says it only searches for a single config file, and that Prettier doesn't support any kind of global configuration.

**Comment by Hans5958** [comment]
> 
> 
> > Yes, but after searching for the config by going up the file tree. This may causes unwanted issue for people that developing with different enviroments, like me.
> 
> Are you sure that's how it works? The Prettier documentation doesn't seem to say anything about that. It explicitly says it only searches for a single config file, and that Prettier doesn't support any kind of global configuration.

There's this paragraph on the page.

> The configuration file will be resolved starting from the location of the file being formatted, and searching up the file tree until a config file is (or isn’t) found.

https://prettier.io/docs/en/configuration.html

Also, some paragraphs on [prettier/prettier-vscode](https://github.com/prettier/prettier-vscode).

> There are multiple options for configuring Prettier with this extension. You can use VS Code settings, prettier configuration files, or an .editorconfig file. The VS Code settings are meant to be used as a fallback and are generally intended only for use on non-project files. **It is recommended that you always include a prettier configuration file in your project specifying all settings for your project.** This will ensure that no matter how you run prettier - from this extension, from the CLI, or from another IDE with Prettier, the same settings will get applied.

https://github.com/prettier/prettier-vscode#configuration

> All prettier options can be configured directly in this extension. These settings are used as a fallback when no configuration file is present in your project, see the configuration section of this document for more details. For reference on the options see the prettier documentation.

https://github.com/prettier/prettier-vscode#prettier-settings

**Comment by WorldLanguages** [comment]
Hmmm... that doesn't say anything about going up the tree for each setting. It says it finds a config file. And Prettier docs say every setting has a default value. If it's missing from the config file, it fallbacks to the Prettier default. So I don't see the issue, making it explicit in the config file would change nothing. What am I missing?

**Comment by Hans5958** [comment]
> Hmmm... that doesn't say anything about going up the tree for each setting. It says it finds a config file.

bruh

> The configuration file will be resolved starting from the location of the file being formatted, and **searching up the file tree** until a config file is (or isn’t) found.

----

> I don't see the issue, making it explicit in the config file would change nothing. What am I missing?

Yes, it won't, but it's just for making sure that Prettier uses the settings that you wanted (in this case, the default settings), and not something else.

**Comment by WorldLanguages** [comment]
I'm closing this, I think Hans finally understood my point lol
